### ISSUE TITLE
Disable strict-aliasing temporarily with GCC for `container_init` in `init_kernel.hpp`

### ISSUE THREAD

**Issue by Alexhuszagh** [libdynd/libdynd#1326]
GCC is fairly well known to be very aggressive with strict aliasing, for good reason. However, on Fedora MinGW (version 7.1.0 20170502), both 32-bit and 64-bit, the `reinterpret_cast<kernel *>` calls trigger `warning: dereferencing type-punned pointer will break strict-aliasing rules`. I believe this is a compiler bug, and not actually a bug with the DyND, but the following does "patch" the issue. 

I'm not sure how desirable this "patch" is, however, implementing it allows Fedora MinGW to fully compile DyND (along with the previous patches mentioned).

```diff
diff --git a/include/dynd/kernels/init_kernel.hpp b/include/dynd/kernels/init_kernel.hpp
index 33fd5af47..83988331c 100644
--- a/include/dynd/kernels/init_kernel.hpp
+++ b/include/dynd/kernels/init_kernel.hpp
@@ -263,6 +263,8 @@ namespace nd {
     void init(const ndt::type &tp, const char *metadata) {
       typedef detail::init_kernel<ResType, ContainerType> kernel;
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstrict-aliasing"
       new (&child) kernel(tp, metadata);
       destruct_wrapper = [](container_init *self) { reinterpret_cast<kernel *>(&self->child)->~kernel(); };
       single_wrapper = [](container_init *self, char *data, const ContainerType &values) {
@@ -271,6 +273,7 @@ namespace nd {
       contiguous_wrapper = [](container_init *self, char *data, const ContainerType *values, size_t size) {
         reinterpret_cast<kernel *>(&self->child)->contiguous(data, values, size);
       };
+#pragma GCC diagnostic pop
     }
 
     container_init(const ndt::type &tp, const char *metadata) {
```

**Comment by insertinterestingnamehere** [comment]
Yep. That's the right way to locally silence a warning like this. This came up in https://github.com/libdynd/libdynd/issues/1315 as well. The solution we agreed on there was to just cut out the warning-is-an-error behavior until we can pin down definitively whether or not this is a false positive. The fact that it's complaining when we use reinterpret_cast is odd since we're already saying "trust us that the types work fine on this one." Silencing it in the mean time is fine with me.

**Comment by Alexhuszagh** [comment]
@insertinterestingnamehere Just an FYI, I'm currently writing my thesis so I will try to get a PR done this week, but if there's been any major delays, I apologize. Feel free to commit changes without me.

**Comment by insertinterestingnamehere** [comment]
Thanks for the heads up. There's no rush. Writing a thesis will keep you very busy. Good luck. I'm working on solving some other issues here at the moment but can come back to these if you don't get to them.
